### PR TITLE
[commissioner] add resign

### PR DESCRIPTION
**Pull request by gjc13**
Add resign to allow other commissioners to petition in deconstruction.
Also made `CommissionerState` an `enum class` to prevent accidental misuse.

### CODE REVIEW

File: src/commissioner/commissioner.hpp [hunk 1]
```diff
@@ -150,6 +150,17 @@ class Commissioner
...
 
+    /**
+     * This method gets number of nodes joined through commissioner
+     *
+     * @returns number of JOIN_FIN.rsp messages sent
+     *
+     */
+    int GetNumFinalizedJoiners(void);
+
+    ~Commissioner();
```
Comment (bukepo) [inline src/commissioner/commissioner.hpp]: add `void`?

File: src/commissioner/commissioner.hpp [hunk 2]
```diff
@@ -159,17 +170,12 @@ class Commissioner
     void CommissionerSet(const SteeringData &aSteeringData);
 
     /**
-     * This method gets number of nodes joined through commissioner
-     *
-     * @returns number of JOIN_FIN.rsp messages sent
+     * This method gracefully resign as commissioner
```
Comment (bukepo) [inline src/commissioner/commissioner.hpp]: resign -> resigns?

File: src/commissioner/commissioner.hpp [hunk 3]
```diff
@@ -181,7 +187,7 @@ class Commissioner
     Commissioner &operator=(const Commissioner &);
 
     int  DtlsHandShake(const sockaddr_in &aAgentAddr);
-    void CommissionerKeepAlive(void);
+    void CommissionerKeepAlive(uint8_t aState);
```
Comment (bukepo) [inline src/commissioner/commissioner.hpp]: uint8_t -> int8_t

File: src/commissioner/commissioner.cpp [hunk 4]
```diff
@@ -101,6 +101,7 @@ static int DummyKeyExport(void *               aContext,
 
 Commissioner::Commissioner(const uint8_t *aPskcBin, int aKeepAliveRate)
     : mDtlsInitDone(false)
+    , mIsCommissioner(false)
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: is this equal to `mCommissionState == CommissionState::kStateAccepted`?

File: src/commissioner/commissioner.cpp [hunk 5]
```diff
@@ -354,13 +355,14 @@ void Commissioner::HandleCommissionerSet(const Coap::Message &aMessage, void *aC
             switch (state)
             {
             case Meshcop::kStateAccepted:
-                commissioner->mCommissionState = kStateAccepted;
+                commissioner->mCommissionState = CommissionState::kStateAccepted;
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: I don't think we should set `mCommissionState` here. This only rejects or accepts the set data.

File: src/commissioner/commissioner.hpp [hunk 6]
```diff
@@ -150,6 +150,17 @@ class Commissioner
      */
     void CommissionerPetition(void);
 
+    /**
+     * This method gets number of nodes joined through commissioner
+     *
+     * @returns number of JOIN_FIN.rsp messages sent
+     *
+     */
+    int GetNumFinalizedJoiners(void);
```
Comment (bukepo) [inline src/commissioner/commissioner.hpp]: `const`?

File: src/commissioner/commissioner.hpp [hunk 7]
```diff
@@ -159,17 +170,12 @@ class Commissioner
...
      *
      */
-    int GetNumFinalizedJoiners(void);
+    void Resign(void);
 
-    ~Commissioner();
-
-private:
-    enum CommissionState
+    enum class CommissionState
```
Comment (jwhui) [inline src/commissioner/commissioner.hpp]: nit: `CommissionState` -> `CommissionerState`

File: src/common/tlv.hpp [hunk 8]
```diff
@@ -148,6 +148,12 @@ class Tlv
         *static_cast<uint8_t *>(GetValue()) = aValue;
     }
 
+    void SetValue(int8_t aValue)
```
Comment (bukepo) [inline src/common/tlv.hpp]: Missing doc string.

File: src/commissioner/commissioner.cpp [hunk 9]
```diff
@@ -450,14 +449,22 @@ void Commissioner::Process(const fd_set &aReadFdSet, const fd_set &aWriteFdSet,
...
-void Commissioner::CommissionerKeepAlive(void)
+void Commissioner::Resign(void)
+{
+    if (mCommissionState == CommissionState::kStateAccepted)
+    {
+        CommissionerKeepAlive(static_cast<int8_t>(Meshcop::kStateRejected));
+    }
+}
+
+void Commissioner::CommissionerKeepAlive(int8_t aState)
```
Comment (jwhui) [inline src/commissioner/commissioner.cpp]: nit: `CommissionerKeepAlive()` -> `SendCommissionerKeepAlive()`

File: src/commissioner/commissioner.cpp [hunk 10]
```diff
@@ -354,13 +354,12 @@ void Commissioner::HandleCommissionerSet(const Coap::Message &aMessage, void *aC
             switch (state)
             {
             case Meshcop::kStateAccepted:
-                commissioner->mCommissionState = kStateAccepted;
                 break;
             case Meshcop::kStateRejected:
-                commissioner->mCommissionState = kStateRejected;
+                commissioner->mCommissionState = CommissionState::kStateRejected;
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: This should not change the commissioner state. It only rejects the set request. The commissioner should continue to be active.

File: src/commissioner/commissioner.cpp [hunk 11]
```diff
@@ -354,13 +354,12 @@ void Commissioner::HandleCommissionerSet(const Coap::Message &aMessage, void *aC
...
                 break;
             default:
-                commissioner->mCommissionState = kStateInvalid;
+                commissioner->mCommissionState = CommissionState::kStateInvalid;
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: Maybe we should ignore this.

File: src/common/tlv.hpp [hunk 12]
```diff
@@ -148,6 +148,15 @@ class Tlv
         *static_cast<uint8_t *>(GetValue()) = aValue;
     }
 
+    /**
+     * This method sets int8_t as the value.
+     */
```
Comment (jwhui) [inline src/common/tlv.hpp]: Add doc for `aValue` parameter.

**Pull request merged**